File: Recursion/is_array_sorted.go
// Given an array, check whether the array is in sorted order with recursion.

package main

import "fmt"

// Time Complexity: O(n). Space Complexity: O(n) for recursive stack space.
func isSorted(A []int) bool {
	n := len(A)
	// Base case 1
	if n == 1 {
		return true
	}
	// Base case 2
	if A[n - 1] < A[n - 2] {
		return false
	}
	// recursive case
	return isSorted(A[:n-1])
}

func main() {
	A := []int{10, 20, 23, 23, 45, 78, 88}
	fmt.Println(isSorted(A))
	A = []int{10, 20, 23, 7, 23, 45, 78, 88}
	fmt.Println(isSorted(A))
}

File: Recursion/is_array_sorted.cpp
// Given an array, check whether the array is in sorted order with recursion.
// Sample Input: [1, 2, 3, 4, 4, 5, 3]
// Output: 0 meaning false
// Sample Output: [1, 2, 3, 4, 4, 5, 6]
// Output: 1 meaning true
#include<bits/stdc++.h>
using namespace std;
bool is_sorted(int A[], int m){
    if(m == 0 || m == 1)
        return true;       
    if(A[0] > A[1])
        return false;
    bool small_array = is_sorted(A + 1, m - 1);
    return small_array;        
}
int main(){
    int A[7] = {1, 2, 3, 4, 4, 5, 3};
    cout << is_sorted(A, 7);
    return 0;
}